bee-ternary/src/convert.rs

```diff
@@ -0,0 +1,369 @@
+    } else {
+        let mut acc = I::zero();
+
+        for trit in trits.iter().rev() {
+            let old_acc = acc.clone();
+            acc = match trit {
+                Btrit::NegOne => -I::one(),
+                Btrit::Zero => I::zero(),
+                Btrit::PlusOne => I::one(),
+            }
+            .checked_add(&old_acc)
+            .and_then(|acc| acc.checked_add(&old_acc))
+            .and_then(|acc| acc.checked_add(&old_acc))
+            .ok_or(Error::TooBig)?;
+        }
+
+        Ok(acc)
+    }
+}
+
+/// Attempt to convert the given trit slice into a number. If the numeric representation of the
+/// trit slice is too large to fit the desired integer, or does not contain any trits, an error
+/// will be returned.
+pub fn trits_to_unsigned_int<I, T: RawEncoding<Trit = Utrit> + ?Sized>(trits: &Trits<T>) -> Result<I, Error>
+where
+    I: Clone + CheckedAdd + Num,
+{
+    if trits.is_empty() {
+        Err(Error::Empty)
+    } else {
+        let mut acc = I::zero();
+
+        for trit in trits.iter().rev() {
+            let old_acc = acc.clone();
+            acc = match trit {
+                Utrit::Zero => I::zero(),
+                Utrit::One => I::one(),
+                Utrit::Two => I::one() + I::one(),
+            }
+            .checked_add(&old_acc)
+            .and_then(|acc| acc.checked_add(&old_acc))
+            .and_then(|acc| acc.checked_add(&old_acc))
+            .ok_or(Error::TooBig)?;
+        }
+
+        Ok(acc)
+    }
+}
+
+/// Produce an iterator over the [`Btrit`]s that make up a given integer.
+pub fn signed_int_trits<I>(x: I) -> impl Iterator<Item = Btrit> + Clone
+where
+    I: Clone + AsPrimitive<i8> + FromPrimitive + Signed,
+{
+    let is_neg = x.is_negative();
+    let mut x = if is_neg { x } else { -x };
+
+    let radix = I::from_i8(3).unwrap();
+
+    std::iter::from_fn(move || {
+        if x.is_zero() {
+            None
+        } else {
+            let modulus = ((x + I::one()).abs() % radix).as_();
+            x = x / radix;
+            if modulus == 1 {
+                x = x + -I::one();
+            }
+            Some(Btrit::try_from(((modulus + 2) % 3 - 1) * if is_neg { -1 } else { 1 }).unwrap())
+        }
+    })
+    // If the integer is exactly 0, add an extra trit
+    .chain(Some(Btrit::Zero).filter(|_| x.is_zero()))
+}
+
+/// Produce an iterator over the [`Utrit`]s that make up a given integer.
+pub fn unsigned_int_trits<I>(mut x: I) -> impl Iterator<Item = Utrit> + Clone
+where
+    I: Clone + AsPrimitive<u8> + FromPrimitive + Num,
+{
+    let radix = I::from_u8(3).unwrap();
+
+    std::iter::from_fn(move || {
+        if x.is_zero() {
+            None
+        } else {
+            let modulus = (x % radix).as_();
+            x = x / radix;
+            Some(Utrit::try_from(modulus).unwrap())
+        }
+    })
+    // If the integer is exactly 0, add an extra trit
+    .chain(Some(Utrit::Zero).filter(|_| x.is_zero()))
+}
+
+#[cfg(test)]
```

Comment:
Should the tests be in `tests` for consistency ?

Reply:
Probably, yes. Will change this.